### ISSUE TITLE
Startup lifecycle events firing multiple times when combined with spring-cloud-starter-config

### ISSUE THREAD

**Issue by edeandrea** [spring-projects/spring-boot#14656]
Spring Boot 2.0.5

[As discussed on StackOverflow](https://stackoverflow.com/questions/52516720/best-way-in-custom-spring-boot-starter-library-to-read-current-properties-and-se) @wilkinsona recommended I file a bug report with a sample reference application.

You can find the sample reference application at https://github.com/edeandrea/test-event-startup

**Comment by spencergibb** [comment]
It's intentional in spring-cloud-commons. Bootstrap creates a parent context.

**Comment by edeandrea** [comment]
There's the set of lifecycle events for the bootstrap (notice in the log the events prior to the "Spring" logo). 

After that there are multiple events of the same type that are even the same instance of the same events that are seen multiple times.

**Comment by spencergibb** [comment]
I don't think it's being "fired" twice, just listened for twice

**Comment by edeandrea** [comment]
If I only have a single event listener should that listener receive the same instance of the same event multiple times? If so then wouldn't a listener need to remember if it had already handled an event?

**Comment by spencergibb** [comment]
I can verify that `AppListener` is getting created twice: once in bootstrap, once in child.

**Comment by edeandrea** [comment]
You are correct. But then the instance in the main context for the same ApplicationPreparedEvent is received 3 times, the same ApplicationReady event is received twice, etc.

**Comment by spencergibb** [comment]
I believe application events are propagated in the context hierarchy (can't remember which direction).

**Comment by edeandrea** [comment]
So then why would I get 2 of one kind of event, 3 of others?

**Comment by spencergibb** [comment]
Not sure. I definitely don't see this as a boot issue. It's come up before in spring cloud. Generally, you can turn off bootstrap or do a guard on the event.

**Comment by edeandrea** [comment]
I think you're right about it not being a boot issue. If I remove the dependency on the config client everything is fine - I only see 1 event of each type (and only a single context).

It just seems odd that just adding a dependency on config client that all of a sudden I need to write defensive logic to make sure I don't process the same event multiple times (once for each context is expected).

Maybe this is an issue in config client?

**Comment by spencergibb** [comment]
It's not config client, it's spring-cloud-commons project, spring-cloud-context module. It's what makes refresh scope work and allows remote configuration to be available before the regular application context starts. As I said, you can disable bootstrap (but you lose features) with `spring.cloud.bootstrap.enabled=false`. It's come up a few times, but generally being defensive is sufficient. We should add a note about how to be defensive with application listeners here http://cloud.spring.io/spring-cloud-static/Finchley.SR1/single/spring-cloud.html#_spring_cloud_context_application_context_services.

**Comment by wilkinsona** [comment]
As discussed on Stack Overflow. That's a Spring Cloud question. They're [on Gitter](https://gitter.im/spring-cloud/spring-cloud).

**Comment by edeandrea** [comment]
Sorry - guess I figured that since there was this discussion here I could add to it. I will do that. Sorry for the inconvenience.
